artiq/coredevice/coredevice_generic.schema.json

```diff
@@ -179,6 +179,89 @@
+                            "maxItems": 1
+                        },
+                        "spi": {
+                            "type": "array",
+                            "items": {
+                                "type": "object",
+                                "properties": {
+                                    "name": {
+                                        "type": "string"
+                                    },
+                                    "clk": {
+                                        "type": "integer",
+                                        "minimum": 0,
+                                        "maximum": 7
+                                    },
+                                    "mosi": {
+                                        "type": "integer",
+                                        "minimum": 0,
+                                        "maximum": 7
+                                    },
+                                    "miso": {
+                                        "type": "integer",
+                                        "minimum": 0,
+                                        "maximum": 7
+                                    },
+                                    "cs": {
+                                        "type": "array",
+                                        "items": {
+                                            "type": "integer",
+                                            "minimum": 0,
+                                            "maximum": 7
+                                        }
+                                    }
+                                },
+                                "required": ["clk"]
+                            },
+                            "minItems": 1
+                        },
+                        "ttl": {
+                            "type": "array",
+                            "items": {
+                                "type": "object",
+                                "properties": {
+                                    "name": {
+                                        "type": "string"
+                                    },
+                                    "pin": {
+                                        "type": "integer",
+                                        "minimum": 0,
+                                        "maximum": 7
+                                    },
+                                    "direction": {
+                                        "type": "string",
+                                        "enum": ["input", "output"]
+                                    },
+                                    "edge_counter": {
+                                        "type": "boolean",
+                                        "default": false
+                                    }
+                                },
+                                "required": ["pin", "direction"]
```

Comment:
The DIO_LVDS EEM can switch direction per I/O pair. Per-TTL direction allows use for this feature. This is important to have inputs and outputs combined on one RJ45 connector. As you probably know, I envision to allow SPI channels, TTL inputs, and TTL outputs combined in a single RJ45 connector. This makes it possible to control external equipment with SPI data, busy signals, and trigger signals with one RJ45 connection.

For other DIO EEM cards (e.g. DIO_SMA) direction can only be set per group of 4. So the directions in the json would have to be configured properly as all DIO cards use the same gateware.

Reply:
So you're suggesting routing single-ended TTL signals as well as SPI signals over the 8 wires of a single RJ45 jack on a DIO_RJ45 (e.g. 4 for SPI standard signals, 4 for TTL signals?). I wasn't aware that was a plan, it seems reasonable & efficient, though it might lose some of the properties of what makes LVDS signaling desirable.

Reply:
No, the DIO_RJ45 only allows differential signals. The 4 pairs could for example be clk, mosi, trigger (TTL out) and busy (TTL in), combining all three in one RJ45.